CorfuDB/CorfuDB
# LR StreamListener that listens across namespaces

## Overview
Currently, stream subscription can be done on a single namespace.  This
PR introduces a special LR-specific streaming subscription to receive
ordered updates across namespaces.

Why should this be merged: Required for Log Replication V2 usecases.

Related issue(s) (if applicable): #<number>

### Testing
**Integration Tests**
Test the first 4 workflows with stream subscription done 1)when in log entry sync and 2)when in snapshot sync: 
- Test streaming updates received on the special LR Multi Namespace Listener.  Verify the order and actual data.
- Test the above with different Tx batch size
- Test the above with a custom buffer size of the LR Streaming Task
- Subscribe the special listener at a timestamp after data is written.  Write more data after subscription.  Verify that data written both before and after subscription is available through mergeTableOnSubscription() and streaming updates.
- Test with concurrent updates to the data and LR status tables after subscription.  Simulates data updates being received within and outside a Snapshot sync.
- Test that updates from non-subscribed tables are not received
- Regression Test: Test the coexistence of an LR and non-LR streaming task and verify that both work as expected.

**Unit Tests**
LRStreampPollingSchedulerTest (extension of StreamingPollingSchedulerTest)
- Test addition of an LR Streaming Task by StreamPollingScheduler.
- Test that a duplicate addition of an LR task with the same listener has the same behavior as a regular streaming task
- Test removal of an LR Task
- Verify that a trimmed exception is thrown when maxAddressSeen < trimMark
- Verify that a trimmed exception is thrown when reading an address which was trimmed
- Verify that LRDeltaStream is polled only when availableSpace >= stream polling threshold

LRStreamingTaskTest (extension of StreamingTaskTest)
- Verify that only valid transitions of the task state are allowed.
- Verify that a hole is not propagated to the client.  Other(data) writes are propagated.

LRDeltaStreamTest (extension of LRStreamingTaskTest)
- Verify the hasNext() and next() apis.  hasNext() returns false when no data to consume.  Returns true on a trim gap.  next() throws a trimmed exception if a trimmed address is read.
- Verify that the LRDelta stream does not add new addresses once it detects a trimMark > lastAddressRead
- Verify that a duplicated/redundant refresh of the stream address space is a no-op
- Verify that the buffer maintained in DeltaStream is refreshed as space it gets consumed(read)
- Verify that updates from a stream not tracked by LRDeltaStream is not expected and leads to an exception
- Start a producer and consumer thread which run concurrently to fill and consume from the LR Delta Stream.  Verify that all expected updates are seen by the consumer.

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc
